src/puya/awst/nodes.py

```diff
@@ -803,24 +803,31 @@ def accept(self, visitor: ExpressionVisitor[T]) -> T:
         return visitor.visit_box_value_expression(self)
 
 
-@attrs.frozen(init=False, eq=False, hash=False)  # use identity equality
+@attrs.frozen
 class SingleEvaluation(Expression):
-    """caveat emptor"""
+    """
+    This node wraps an underlying expression and effectively caches the result of that lowering,
+    such that regardless of how many times the SingleEvaluation object
+    (or any object comparing equal to it) appears in the AWST,
+    the underlying source expression will only be evaluated once.
+    """
 
     source: Expression
+    _id: int = attrs.field()
```

Comment:
Does this need to be public for serialization to set it?

Reply:
nah, nothing is ever truly private in Python :P 
`cattrs` handles it fine.